### PR TITLE
[QoI] Improve diagnostic for raw representable argument mismatch

### PR DESCRIPTION
**Pull request by xedin**
Covers some common situations when switching from literal types e.g.
string or int to more Swift Way of raw representable enum of choices.

Resolves: rdar://problem/32431165, rdar://problem/31977537, rdar://problem/32432253

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

File: test/Sema/enum_raw_representable.swift [hunk 1]
```diff
@@ -103,3 +103,40 @@ func rdar32431736() {
...
+
+func rdar32431165_1(_: E_32431165) {}
+func rdar32431165_1(_: Int) {}
+func rdar32431165_1(_: Int, _: E_32431165) {}
+
+rdar32431165_1("")
+// expected-error@-1 {{cannot convert value of type 'String' to expected argument type 'E_32431165'}} {{15-15=E_32431165(rawValue: }} {{19-19=)}}
+rdar32431165_1(42, "")
+// expected-error@-1 {{cannot convert value of type 'String' to expected argument type 'E_32431165'}} {{20-20=E_32431165(rawValue: }} {{22-22=)}}
+
```
Comment (rudkx) [inline test/Sema/enum_raw_representable.swift]: Can you add a test for a nonexistent enum element?
```
_ = rdar32431165_1(.nada)
```

File: lib/Sema/CSDiag.cpp [hunk 2]
```diff
@@ -5317,6 +5317,96 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
...
+    case CC_OneGenericArgumentMismatch:
+    case CC_ArgumentNearMismatch:
+    case CC_ArgumentMismatch:
+      break;
+
+    default:
+      return false;
+  }
+
+  auto argType = argExpr->getType();
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: Can you use `ConstraintSystem::getType(argExpr)` here?

I'm inching closer to getting all the CS code switched over to the side map but new appearances of `Expr::getType()` keep creeping in.

File: lib/Sema/CSDiag.cpp [hunk 3]
```diff
@@ -5317,6 +5317,96 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
...
+  if (!argType || argType->hasTypeVariable() || argType->hasUnresolvedType())
+    return false;
+
+  ArrayRef<KnownProtocolKind> rawRepresentableProtocols = {
+      KnownProtocolKind::ExpressibleByStringLiteral,
+      KnownProtocolKind::ExpressibleByIntegerLiteral};
+
+  const auto *CS = CCI.CS;
+  auto arguments = decomposeArgType(argExpr->getType(), argLabels);
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: Same re: `getType()`.

File: lib/Sema/CSDiag.cpp [hunk 4]
```diff
@@ -5317,6 +5317,96 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
...
+      continue;
+
+    for (unsigned i = 0, n = parameters.size(); i != n; ++i) {
+      auto paramType = parameters[i].Ty;
+      auto argType = arguments[i].Ty;
+
+      for (auto kind : rawRepresentableProtocols) {
+        // If trying to convert from raw type to raw representable,
+        // or vice versa from raw representable (e.g. enum) to raw type.
+        if (isRawRepresentableMismatch(argType, paramType, kind, CS)) {
```
Comment (rudkx) [inline lib/Sema/CSDiag.cpp]: Can you change this to:
```
  if (!isRaw...)
    continue;
```

to reduce indentation?

**Review by rudkx — COMMENTED** [review]
LGTM other than the handful of comments.

**Comment by xedin** [comment]
Changed the `if` statement and added couple more tests as requested. Also talked to @rudkx offline and we've decided not to change `argExpr->getType` because there are couple of issues with that related to the expressions which are checked separately, he is going to take a look at the later.

**Comment by xedin** [comment]
@swift-ci please smoke test and merge

**Comment by xedin** [comment]
@swift-ci please smoke test and merge

**Pull request merged**